resources/turso/src/lib.rs

```diff
@@ -0,0 +1,109 @@
+use async_trait::async_trait;
+use libsql_client::{client::Client, new_client_from_config, Config};
+use serde::{Deserialize, Serialize};
+use shuttle_service::{
+    error::{CustomError, Error as ShuttleError},
+    Factory, ResourceBuilder, Type,
+};
+use url::Url;
+
+#[derive(Serialize, Deserialize, Default)]
+pub struct Turso {
+    addr: String,
+    token: String,
+    local_addr: Option<String>,
+}
+
+#[derive(Serialize, Deserialize)]
+pub struct TursoOutput {
+    conn_url: Url,
+    token: Option<String>,
+}
+
+impl Turso {
+    pub fn local_addr(mut self, local_addr: &str) -> Self {
+        self.local_addr = Some(local_addr.to_string());
+        self
+    }
+}
+
+pub enum Error {
+    UrlParseError(url::ParseError),
+}
+
+impl From<Error> for shuttle_service::Error {
+    fn from(error: Error) -> Self {
+        let msg = match error {
+            Error::UrlParseError(err) => format!("Cannot parse Turso Url: {}", err),
+        };
+
+        ShuttleError::Custom(CustomError::msg(msg))
+    }
+}
+
+#[async_trait]
+impl ResourceBuilder<Client> for Turso {
+    const TYPE: Type = Type::Turso;
+
+    type Config = Self;
+    type Output = TursoOutput;
+
+    fn new() -> Self {
+        Self::default()
+    }
+
+    fn config(&self) -> &Self::Config {
+        &self
+    }
+
+    async fn output(
+        self,
+        factory: &mut dyn Factory,
+    ) -> Result<Self::Output, shuttle_service::Error> {
+        match factory.get_environment() {
+            shuttle_service::Environment::Production => {
+                if self.addr.is_empty() {
+                    // XXX: should we raise an error even not running in a production
+                    // environment ?
+                    Err(ShuttleError::Custom(CustomError::msg("missing addr")))
+                } else {
+                    Ok(TursoOutput {
+                        // XXX: should we allow for other kind of connection string ? Not having libsql at
+                        // the start of the url even though the Turso CLI is printing it might be confusing
+                        // instead of just giving the responsability to the user.
+                        conn_url: Url::parse(&format!("libsql://{}", self.addr))
+                            .map_err(Error::UrlParseError)?,
+                        token: Some(self.token),
+                    })
+                }
+            }
+            shuttle_service::Environment::Local => {
+                // Default to a local db of the name of the service.
+                // XXX: should we check that the file doesn't exist before using that name ?
```

Comment:
Does `.get_build_path()` already covert that case? It might lead to confusing error messages if the file doesn't exist or can't be created.

Reply:
`.get_build_path()` covers the case of the folder not exiting which would lead to the errors you mention. The case I was talking about is when inside the folder returned by `.get_build_path()`, what should happen if database file already exists ? This is obviously okay if that's the database that has been created previously, but it might collude with a pre-existing file. Should we instead force the user to define the name of the database to use locally instead of choosing one for them ? Or do we just don't care ?

Reply:
If the file already exists and is not a valid database, the code will end up throwing errors somewhere else. I think we should just make sure we print useful error messages in those cases rather than putting that burden on the user.

Reply:
When you try to read or do anything with the database, the error is `File opened that is not a database file`, which seems fine to me :)